File: shared/tscanner-common/src/types/display.ts
import z from 'zod';
import { severitySchema } from './enums';

const displayIssueSchema = z.object({
  line: z.number(),
  column: z.number(),
  message: z.string(),
  lineText: z.string().optional(),
  ruleName: z.string().optional(),
});

const fileIssuesSchema = z.object({
  filePath: z.string(),
  issues: z.array(displayIssueSchema),
});

const ruleGroupSchema = z.object({
  ruleName: z.string(),
  severity: severitySchema,
  issueCount: z.number(),
  fileCount: z.number(),
  files: z.array(fileIssuesSchema),
});

export type DisplayIssue = z.infer<typeof displayIssueSchema>;
export type FileIssues = z.infer<typeof fileIssuesSchema>;
export type RuleGroup = z.infer<typeof ruleGroupSchema>;

File: packages/rust-core/crates/tscanner_types/src/display.rs
use crate::enums::Severity;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DisplayIssue {
    pub line: usize,
    pub column: usize,
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub line_text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rule_name: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FileIssues {
    pub file_path: String,
    pub issues: Vec<DisplayIssue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuleGroup {
    pub rule_name: String,
    pub severity: Severity,
    pub issue_count: usize,
    pub file_count: usize,
    pub files: Vec<FileIssues>,
}
